invertedai/large/initialize.py

```diff
@@ -0,0 +1,265 @@
+from invertedai.common import TrafficLightStatesDict, Point
+from invertedai.utils import get_default_agent_attributes
+from invertedai.error import InvertedAIError
+
+SLACK = 2
+AGENT_SCOPE_FOV_BUFFER = 20
+
+@validate_call
+def define_regions_grid(
+    map_center: Optional[Tuple[float,float]] = (0.0,0.0),
+    width: Optional[float] = 100.0, 
+    height: Optional[float] = 100.0, 
+    stride: Optional[float] = 50.0
+) -> List[Region]:
+    def check_valid_center(center):
+        return (map_center[0] - width) < center[0] < (map_center[0] + width) and \
+            (map_center[1] - height) < center[1] < (map_center[1] + height)
+
+    def get_neighbors(center):
+        return [
+            (center[0] + (i * stride), center[1] + (j * stride))
+            for i, j in list(product(*[(-1, 1),]* 2))
+        ]
+
+    queue, centers = [map_center], []
+
+    while queue:
+        center = queue.pop(0)
+        neighbors = filter(check_valid_center, get_neighbors(center))
+        queue.extend([
+            neighbor
+            for neighbor in neighbors
+            if neighbor not in queue and neighbor not in centers
+        ])
+        if center not in centers and check_valid_center(center):
+            centers.append(center)
+    
+    regions = [None]*len(centers)
+    for i, center in enumerate(centers):
+        regions[i] = Region.fromlist([Point.fromlist(list(center))])
+
+    return regions
+
+@validate_call
+def get_regions_density_by_road_area(
+    location: str,
+    regions: List[Region],
+    max_agent_density: Optional[int] = 10,
+    scaling_factor: Optional[float] = 1.0,
+    display_progress_bar: Optional[bool] = True
+) -> List[Region]:
+    #Get fraction of image that is a drivable surface (assume all non-black pixels are drivable)
+    center_road_area_dict = {}
+    max_drivable_area_ratio = 0
+    
+    iterable_regions = None
+    if display_progress_bar:
+        iterable_regions = tenumerate(
+            regions, 
+            total=len(regions),
+            desc=f"Calculating drivable surface areas"
+        )
+    else:
+        iterable_regions = regions
+
+    for i, region in iterable_regions:
+        #Naively check every square within requested area
+        center_tuple = (region.center.x, region.center.y)
+        birdview = iai.location_info(
+            location=location,
+            rendering_fov=int(region.get_region_fov()),
+            rendering_center=center_tuple
+        ).birdview_image.decode()
+
+        ## Get number of black pixels
+        birdview_arr_shape = birdview.shape
+        total_num_pixels = birdview_arr_shape[0]*birdview_arr_shape[1]
+        # Convert to grayscale using Luminosity Method: gray = 0.114*B + 0.587*G + 0.299*R
+        # Image should be in BGR color pixel format
+        birdview_grayscale = np.matmul(birdview.reshape(total_num_pixels,3),np.array([[0.114],[0.587],[0.299]]))
+        number_of_black_pix = np.sum(birdview_grayscale == 0)
```

Comment:
```suggestion
        number_of_black_pix = np.sum((birdview.sum(axis=-1) == 0).int())
```
This is a very elaborate conversion only to compare things to zero at the end.

Reply:
Hm yeah I guess at this point we know we don't need to do anything more fancy with a grayscale image. I made this change.